querier/querier.go

```diff
@@ -135,8 +140,39 @@ func (qm MergeQuerier) QueryInstant(ctx context.Context, ts model.Time, stalenes
 
 // MetricsForLabelMatchers Implements local.Querier.
 func (qm MergeQuerier) MetricsForLabelMatchers(ctx context.Context, from, through model.Time, matcherSets ...metric.LabelMatchers) ([]metric.Metric, error) {
-	// TODO: Implement.
-	return nil, nil
+	mss, errs := make(chan []metric.Metric), make(chan error)
+	for _, q := range qm.Queriers {
+		func(q Querier) {
+			ms, err := q.MetricsForLabelMatchers(ctx, from, through, matcherSets...)
+			if err != nil {
+				errs <- err
+			} else {
+				mss <- ms
+			}
+		}(q)
+	}
+
+	metrics := map[model.Fingerprint]metric.Metric{}
+	var lastErr error
+	for range qm.Queriers {
+		select {
+		case ms := <-mss:
+			for _, m := range ms {
+				metrics[m.Metric.FastFingerprint()] = m
```

Comment:
N.B. `Fingerprint()` can theoretically still create collisions when used without a mapper, but in practice with a small set of metrics being fingerprinted at a time, we are ignoring that.